internal/metric/metric.go

```diff
@@ -0,0 +1,196 @@
+		return err
+	}
+	for _, metric := range metrics {
+		if err := encoder.Encode(metric); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Count returns a HandlerFunc that wraps h and counts the
+// how many requests succeeded (HTTP 200 OK) and how many
+// failed.
+//
+// Count distingushes requests that fail with some sort of
+// well-defined error (HTTP 4xx) and requests that fail due
+// to some internal error (HTTP 5xx).
+func (m *Metrics) Count(h http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		h(&countResponseWriter{
+			ResponseWriter: w,
+			flusher:        w.(http.Flusher),
+			succeeded:      m.requestSucceeded,
+			errored:        m.requestErrored,
+			failed:         m.requestFailed,
+		}, r)
+	}
+}
+
+// Latency returns a HandlerFunc that wraps h and measures the
+// internal request-response latency.
+//
+// The internal request-response latency is the time the
+// application takes to generate and send a response after
+// receiving a request. It basically shows how many request
+// the application can handle.
+func (m *Metrics) Latency(h http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		h(&latencyResponseWriter{
+			ResponseWriter: w,
+			flusher:        w.(http.Flusher),
+			start:          time.Now(),
+			histogram:      m.requestLatency,
+		}, r)
+	}
+}
+
+// latencyResponseWriter is an http.ResponseWriter that
+// measures the internal request-response latency.
+type latencyResponseWriter struct {
+	http.ResponseWriter
+	flusher http.Flusher
+
+	start     time.Time            // The point in time when the request was received
+	histogram prometheus.Histogram // The latency histogram
+	written   bool                 // Inidicates whether the HTTP headers have been written
+}
+
+var (
+	_ http.ResponseWriter = (*latencyResponseWriter)(nil)
+	_ http.Flusher        = (*latencyResponseWriter)(nil)
+)
+
+func (w *latencyResponseWriter) WriteHeader(status int) {
+	w.ResponseWriter.WriteHeader(status)
+	if !w.written {
+		w.histogram.Observe(time.Since(w.start).Seconds())
+		w.written = true
+	}
+}
+
+func (w *latencyResponseWriter) Flush() {
+	if w.flusher != nil {
+		w.flusher.Flush()
+	}
+}
+
+// countResponseWriter is an http.ResponseWriter that
+// counts the number of requests partition by requests
+// that:
+//   • Succeeded (HTTP 200 OK)
+//   • Errored   (HTTP 4xx)
+//   • Failed    (HTTP 5xx)
+type countResponseWriter struct {
+	http.ResponseWriter
+	flusher http.Flusher
+
+	succeeded prometheus.Counter
+	errored   prometheus.Counter
+	failed    prometheus.Counter
+	written   bool // Inidicates whether the HTTP headers have been written
+}
+
+var (
+	_ http.ResponseWriter = (*countResponseWriter)(nil)
+	_ http.Flusher        = (*countResponseWriter)(nil)
+)
+
+func (w *countResponseWriter) WriteHeader(status int) {
+	w.ResponseWriter.WriteHeader(status)
+	if !w.written {
+		switch {
+		case status == http.StatusOK:
```

Comment:
```suggestion
		case status == http.StatusOK || status == http.StatusNoContent:
```
should we check for `ok no content` status as well?, there is lot more of ok responses too:

```
StatusOK                   = 200 // RFC 7231, 6.3.1
	StatusCreated              = 201 // RFC 7231, 6.3.2
	StatusAccepted             = 202 // RFC 7231, 6.3.3
	StatusNonAuthoritativeInfo = 203 // RFC 7231, 6.3.4
	StatusNoContent            = 204 // RFC 7231, 6.3.5
	StatusResetContent         = 205 // RFC 7231, 6.3.6
	StatusPartialContent       = 206 // RFC 7233, 4.1
	StatusMultiStatus          = 207 // RFC 4918, 11.1
	StatusAlreadyReported      = 208 // RFC 5842, 7.1
	StatusIMUsed               = 226 // RFC 3229, 10.4.1
```

Reply:
That is correct. However, no `http.HandlerFunc` implemented by KES returns anything else (or is supposed to return anything else)  then `200 OK`. If that should ever change it should be tracked as well - maybe separately. To me e.g. `PartialContent` or `IMUsed` is quite different from `OK`.

If someone writes a`http.HandlerFunc` that returns something other then `200 OK` in case of success but forgets to update the metrics then we fail early:
```
			// We panic to signal that the server returned a status code
			// that is not tracked. If, in the future, the application
			// returns a new (kind of) status code it should be collected
			// as well.
			// Otherwise, we would silently ignore new status codes and the
			// metrics would be incomplete.
			panic("metrics: unexpected response status code " + strconv.Itoa(status))
```
(see a few lines below)